pkg/nats/stream/producer_client.go

```diff
@@ -0,0 +1,223 @@
+package stream
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/nats-io/nats.go"
+	"github.com/rs/zerolog/log"
+)
+
+type ProducerClientParams struct {
+	Conn   *nats.Conn
+	Config StreamProducerClientConfig
+}
+
+type ProducerClient struct {
+	Conn *nats.Conn
+	mu   sync.RWMutex // Protects access to activeStreamInfo and activeConnHeartBeatRequestSubjects
+
+	// A map of ConsumerID to StreamId that are active
+	activeStreamInfo map[string]map[string]StreamInfo
+	// A map of ConsumerID to the subject where a heartBeatRequest needs to be sent.
+	activeConnHeartBeatRequestSubjects map[string]string
+	heartBeatCancelFunc                context.CancelFunc
+	config                             StreamProducerClientConfig
+}
+
+func NewProducerClient(ctx context.Context, params ProducerClientParams) (*ProducerClient, error) {
+	nc := &ProducerClient{
+		Conn:                               params.Conn,
+		activeStreamInfo:                   make(map[string]map[string]StreamInfo),
+		activeConnHeartBeatRequestSubjects: make(map[string]string),
+		config:                             params.Config,
+	}
+
+	go nc.heartBeat(ctx)
```

Comment:
As we've discussed this, I am not on favour of having a dedicated `Start` method for a client instance where uses don't really need to know about background routines. It doesn't feel right to tell users to start a _client_, and we will have to cover edge cases such as multiple starts, starting after stopping, sending/handling requests for unstarted client, and so on. I understand the overall value, specially when you have multiple dependencies that you want to setup first before starting the service, but those don't apply here.

In fact, Udit is planning to optimize things where we only run the background heartbeat routine lazily if and when we have active streams, and stop it when we no longer have streams. Regardless whether we feel this is an over optimization or not, having the background routine to be internal to the client and not exposed to users to start enables these kinds of optimizations and simplifies the developer experience.

What I agree with is avoiding passing the context to the constructor, and using that context to manage the lifecycle of the internal go routines. Contexts are better suited to manage the lifetime of requests, not services or processes. Their limitation is that they don't provide a way to ensure it is safe to exit the service or control the order of service shutdown. It's better to make these services explicitly closable and control when to close them as part of[ node cleanup](https://github.com/bacalhau-project/bacalhau/blob/23a474b8e48af89b97c37827389e100918b624d3/pkg/node/compute.go#L224), or in the future through fx lifecycle management. We should avoid using contexts here. 

This change requires adding a `Close` method from the clients, proxies and handlers. Udit already opened an issue to improve client graceful shutdown, and this can be improved as part of that issue